resilience4j-bulkhead/src/main/java/io/github/resilience4j/bulkhead/adaptive/AdaptiveBulkhead.java

```diff
@@ -0,0 +1,575 @@
+import io.github.resilience4j.core.EventPublisher;
+import io.vavr.CheckedConsumer;
+import io.vavr.CheckedFunction0;
+import io.vavr.CheckedFunction1;
+import io.vavr.CheckedRunnable;
+
+/**
+ * A Bulkhead instance is thread-safe can be used to decorate multiple requests.
+ * <p>
+ * A {@link AdaptiveBulkhead} represent an entity limiting the amount of parallel operations. It does not assume nor does it mandate usage
+ * of any particular concurrency and/or io model. These details are left for the client to manage. This bulkhead, depending on the
+ * underlying concurrency/io model can be used to shed load, and, where it makes sense, limit resource use (i.e. limit amount of
+ * threads/actors involved in a particular flow, etc).
+ * <p>
+ * In order to execute an operation protected by this bulkhead, a permission must be obtained by calling {@link AdaptiveBulkhead#tryAcquirePermission()} ()}
+ * If the bulkhead is full, no additional operations will be permitted to execute until space is available.
+ * <p>
+ * Once the operation is complete, regardless of the result, client needs to call {@link AdaptiveBulkhead#onSuccess(long, TimeUnit)}  in order to maintain
+ * integrity of internal bulkhead state which is handled as the following :
+ * <p>
+ * Adaptive capacity management by default use AIMD algorithm for limit control but the user can inject custom limiter implementation by implementing {@link LimitPolicy}
+ * <p>
+ */
+public interface AdaptiveBulkhead {
+	/**
+	 * Acquires a permission to execute a call, only if one is available at the time of invocation.
+	 *
+	 * @return {@code true} if a permission was acquired and {@code false} otherwise
+	 */
+
+	boolean tryAcquirePermission();
+
+	/**
+	 * Acquires a permission to execute a call, only if one is available at the time of invocation
+	 *
+	 * @throws BulkheadFullException when the Bulkhead is full and no further calls are permitted.
+	 */
+	void acquirePermission();
+
+	/**
+	 * Releases a permission and increases the number of available permits by one.
+	 * <p>
+	 * Should only be used when a permission was acquired but not used. Otherwise use
+	 * {@link AdaptiveBulkhead#onSuccess(long, TimeUnit)} to signal a completed call and release a permission.
+	 */
+	void releasePermission();
+
+	/**
+	 * Records a successful call and releases a permission.
+	 */
+	void onSuccess(long startTime, TimeUnit durationUnit);
+
+	/**
+	 * Records a failed call and releases a permission.
+	 */
+	void onError(long startTime, TimeUnit durationUnit,Throwable throwable);
+
+
+	/**
+	 * Returns the name of this bulkhead.
+	 *
+	 * @return the name of this bulkhead
+	 */
+	String getName();
+
+	/**
+	 * Returns the AdaptiveBulkheadConfig of this Bulkhead.
+	 *
+	 * @return bulkhead config
+	 */
+	AdaptiveBulkheadConfig getBulkheadConfig();
+
+	/**
+	 * Get the AdaptiveMetrics of this Bulkhead.
+	 *
+	 * @return the AdaptiveMetrics of this Bulkhead
+	 */
+	AdaptiveMetrics getMetrics();
```

Comment:
I think we should stick to `AdaptiveBulkead.Metrics` like in our other components. Even if Sonar complains about it ;)